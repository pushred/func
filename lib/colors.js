const chroma = require('chroma-js');
const isPlainObject = require('lodash.isplainobject');

function parse(colors) {
  return Object.keys(colors).reduce((output, colorKey) => {
    let colorVals = colors[colorKey];
    if (typeof colorVals === 'string') output[colorKey] = chroma(colorVals);
    if (!isPlainObject(colorVals)) return output;

    const colorModelName = Object.keys(colorVals).map(val => val[0]).join('');
    colorVals = Object.keys(colorVals).map(key => colorVals[key]);

    output[colorKey] = chroma.call(null, colorVals, colorModelName);

    return output;
  }, {});
}

module.exports = {
  parse,
};
