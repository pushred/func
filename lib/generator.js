const kebabCase = require('lodash.kebabcase');

const { parse } = require('./colors');

/**
 * adjustColor
 * Applies Chroma.js functions
 * @private
 */

function adjustColor(color, adjuster) {
  const name = adjuster.split('(')[0];
  const value = /\((.*)\)/.exec(adjuster)[1];

  const decimalValue = value.includes('%')
    ? parseInt(value, 10) / 100
    : value;

  if (name === 'a' || name === 'alpha' || name === 'opacity') {
    return color.alpha(decimalValue);
  }

  return color;
}

/**
 * getProperty
 * Maps a class containing a shorthand alias to a full CSS property name
 * @private
 */

function getProperty({
  className,
  propertyAliases = {
    bg: 'background-color',
    border: 'border-color',
  },
}) {
  const alias = Object.keys(propertyAliases).find(alias => className.includes(alias));
  if (!alias) return 'color';

  return propertyAliases[alias];
}

/**
 * expandClasses
 * Expands color/pseudo-class shorthand config into one or more classes
 */

function expandClasses({ classes, colors }) {
  const chromaColors = parse(colors);

  return Object.keys(classes).reduce((output, className) => {
    const property = getProperty({ className });

    const value = classes[className];
    const adjusterFn = /[a-z]*\(.*\)/.exec(value);
    const props = adjusterFn
      ? value.replace(adjusterFn[0], '').split(/\s/)
      : value.split(/\s/);

    const states = props
      .filter(name => name.includes(':'))
      .map(state => {
        if (state[0] === '&') return state.substr(1); // nest-style
        if (state[0] === ':') return state; // shorthand
        return ' ' + state; // descendent
      });

    const colorKey = props[0];
    const colorValue = adjusterFn
      ? adjustColor(chromaColors[colorKey], adjusterFn[0])
      : chromaColors[colorKey];

    if (!colorValue) return output;

    const styles = `${property}: ${colorValue.css('hsl').replace(/,/g, ', ')}`;

    return [
      ...output,
      `.${className} { ${styles} }`,
      ...states.map(pseudoClassName => `.${className}${pseudoClassName} { ${styles} }`),
    ];
  }, []).join('\n');
}

/**
 * generateClasses
 * Creates a named color class for each configured property
 */

function generateClasses({ colors, properties = {}, states = [] }) {
  const chromaColors = parse(colors);

  return Object.keys(colors).reduce((output, colorKey) => {
    const colorName = kebabCase(colorKey);
    const colorValue = chromaColors[colorKey];

    return [
      ...output,
      ...Object.keys(properties).map(property => {
        const className = properties[property].replace('${name}', colorName);

        const styles = `${property}: ${colorValue.css('hsl').replace(/,/g, ', ')}`;

        return [
          `.${className} { ${styles} }`,
          ...states.map(pseudoClassName => `.${className}:${pseudoClassName} { ${styles} }`),
        ];
      }),
    ];
  }, []).join('\n');
}

module.exports = {
  expandClasses,
  generateClasses,
};
