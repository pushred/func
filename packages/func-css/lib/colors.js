const chroma = require('chroma-js');
const isPlainObject = require('lodash.isplainobject');

const MIXTURES = ['shades', 'tints'];

const STOP_NAMES = {
  shade: [
    'dark',
    'darker',
    'darkest',
    'x-dark',
    'xx-dark',
    'xxx-dark',
  ],
  tint: [
    'light',
    'lighter',
    'lightest',
    'x-light',
    'xx-light',
    'xxx-light',
  ],
};

/**
 * @typedef {Object} Color
 * chroma.js color instance
 */

/**
 * parse
 * Parses color strings and configs to chroma.js instances
 * @private
 */

function parse(colors = {}) {
  return Object.keys(colors).reduce((output, colorKey) => {
    const rawColor = colors[colorKey];

    if (typeof rawColor === 'string') {
      output[colorKey] = getColorData({
        color: chroma(rawColor),
        colorName: colorKey,
      });
    }

    if (!isPlainObject(rawColor)) return output;

    const config = Object.keys(rawColor).reduce((config, key) => {
      (MIXTURES.includes(key))
        ? config.mixtures[key] = rawColor[key]
        : config.color[key] = rawColor[key];
      return config;
    }, {
      color: {},
      mixtures: {},
    });

    let colorModelName = Object.keys(config.color).map(val => val[0]).join('');
    if (colorModelName === 'hsb') colorModelName = 'hsv';

    const colorSpec = Object.keys(config.color).map(key => {
      const originalValue = config.color[key];
      const isPctValue = ['s', 'b', 'l', 'v'].includes(key[0]);
      if (!isPctValue) return originalValue;

      const value = parseFloat(originalValue);

      if (value === 0) return value;
      if (value > 1) return value / 100;
      return originalValue;
    });

    const hasHue = colorModelName === 'hsl' || colorModelName === 'hsv';
    const isBlack = colorSpec[1] === 0 && colorSpec[2] === 0;
    const isWhite = colorSpec[1] === 0 && colorSpec[2] === 1;

    if (hasHue && colorSpec[0] === 0 && (isBlack || isWhite)) {
      throw new Error('`0` is not a valid hue, pure black/white have none');
    }

    const baseColor = chroma.call(null, colorSpec, colorModelName);

    return {
      ...output,
      [colorKey]: getColorData({ color: baseColor, colorName: colorKey }),
      ...Object.keys(config.mixtures).reduce((mixtures, mixture) => ({
        ...mixtures,
        ...getMixtures({
          baseColor,
          colorName: colorKey,
          method: mixture.slice(0, -1), // singular
          stops: config.mixtures[mixture].stops,
        }),
      }), {}),
    };
  }, {});
}

/**
 * getColorData
 * Expands parsed color metadata
 * @param {String} colorName
 * @param {Color} color
 * @private
 */

 function getColorData({ color, colorName }) {
  return {
    color,
    colorName,
    hex: color.hex('rgb'),
    hsl: color.css('hsl'),
    rgb: color.css('rgb'),
  };
 }

/**
 * getMixtures
 * Generates a set of mixtures
 * @param {Color} baseColor
 * @param {('shade'|'tint')} method
 * @param {Number} stops
 * @private
 */

function getMixtures({ baseColor, colorName, method, stops = 2 } = {}) {
  if (!baseColor) return console.error('baseColor required');
  if (!method) return console.error('method required');

  let mixColor;
  if (method === 'tint') mixColor = chroma('white');
  if (method === 'shade') mixColor = chroma('black');

  return mix({
    stops,
    color1: baseColor,
    color2: mixColor,
  }).reduce((values, value, index) => ({
    ...values,
    [`${STOP_NAMES[method][index]}-${colorName}`]: {
      ...getColorData({ color: value, colorName }),
      mixture: {
        method,
        stop: STOP_NAMES[method][index],
      },
    },
  }), {});
}

/**
 * mixColor
 * Creates color mixtures for a specified number of stops
 * @param {Color} color1
 * @param {Color} color2
 * @param {Number} stops
 * @private
 */

function mix({ color1, color2, stops = 2 } = {}) {
  const ratioIncrement = 1 / (stops + 1);
  return [...Array(stops)].map((stop, index) => {
    return chroma.mix(color1, color2, ratioIncrement * (index + 1))
  });
}

module.exports = {
  mix,
  parse,
};
