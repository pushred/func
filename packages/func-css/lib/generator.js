const camelCase = require('lodash.camelcase');
const kebabCase = require('lodash.kebabcase');

const { parse } = require('./colors');

/**
 * adjustColor
 * Applies Chroma.js functions
 * @private
 */

function adjustColor(color, adjuster) {
  const name = adjuster.split('(')[0];
  const value = /\((.*)\)/.exec(adjuster)[1];

  const decimalValue = value.includes('%')
    ? parseInt(value, 10) / 100
    : value;

  if (name === 'a' || name === 'alpha' || name === 'opacity') {
    return color.alpha(decimalValue);
  }

  return color;
}

/**
 * getProperty
 * Maps a class containing a shorthand alias to a full CSS property name
 * @private
 */

function getProperty({
  className,
  propertyAliases = {
    bg: 'background-color',
    border: 'border-color',
  },
}) {
  const alias = Object.keys(propertyAliases).find(alias => className.includes(alias));
  if (!alias) return 'color';

  return propertyAliases[alias];
}

/**
 * expandClasses
 * Expands color/pseudo-class shorthand config into one or more classes
 */

function expandClasses({ classes = {}, colors = {} }) {
  return Object.keys(classes).reduce((output, className) => {
    const property = getProperty({ className });

    const value = classes[className];
    const adjusterFn = /[a-z]*\(.*\)/.exec(value);

    const props = adjusterFn
      ? value.replace(adjusterFn[0], '').split(/\s/)
      : value.split(/\s/);

    const states = props
      .filter(name => name.includes(':'))
      .map(state => {
        if (state[0] === '&') return state.substr(1); // nest-style
        if (state[0] === ':') return state; // shorthand
        return ' ' + state; // descendent
      });

    const colorKey = props[0];
    const colorData = colors[colorKey];

    // assume any values missing from palette are CSS keywords that should be used as-is
    let colorValue = colorKey;
    if (colorData && !adjusterFn) colorValue = colorData.color.css('hsl');
    if (adjusterFn) colorValue = adjustColor(colorData.color, adjusterFn[0]).css('hsl');

    const styles = `${property}: ${colorValue.replace(/,/g, ', ')}`;
    const hasInteractState = value.includes('focus') || value.includes('hover');

    return [
      ...output,
      !hasInteractState && `.${className} { ${styles} }`,
      ...states.map(pseudoClassName => `.${className}${pseudoClassName} { ${styles} }`),
    ].filter(Boolean);
  }, []).join('\n');
}

/**
 * generateClasses
 * Creates a named color class for each configured property
 */

function generateClasses({ colors = {}, properties = {}, states = [] }) {
  return Object.keys(colors).reduce((output, colorKey) => {
    const colorName = kebabCase(colorKey);
    const colorValue = colors[colorKey];

    return [
      ...output,
      ...Object.keys(properties).map(property => {
        const className = properties[property].replace('${name}', colorName);

        const styles = `${property}: ${colorValue.color.css('hsl').replace(/,/g, ', ')}`;

        return [
          `.${className} { ${styles} }`,
          ...states.map(pseudoClassName => `.${className}:${pseudoClassName} { ${styles} }`),
        ].join('\n');
      }),
    ];
  }, []).join('\n');
}

/**
 * generateProps
 * Create an object of normalized hex color properties for each class and color
 */

function generateProps({ classes = {}, colors = {} }) {
  const classProps = Object.keys(classes).reduce((output, className) => {
    const value = classes[className];
    const adjusterFn = /[a-z]*\(.*\)/.exec(value);
    const props = adjusterFn
      ? value.replace(adjusterFn[0], '').split(/\s/)
      : value.split(/\s/);

    const colorKey = props[0];
    const colorData = colors[colorKey];

    if (colorData === undefined) return output;

    const chromaColor = adjusterFn
      ? adjustColor(colorData.color, adjusterFn[0])
      : colorData.color;

    return {
      ...output,
      [camelCase(className)]: chromaColor.alpha() === 1
        ? chromaColor.hex()
        : chromaColor.css('rgba').replace(/,/g, ', ')
    };
  }, {});

  const colorProps = Object.keys(colors).reduce((output, colorName) => {
    if (!colors[colorName]) return output;
    return {
      ...output,
      [camelCase(colorName)]: colors[colorName].color.hex()
    };
  }, {});

  return {
    ...classProps,
    ...colorProps,
  };
}

module.exports = {
  expandClasses,
  generateClasses,
  generateProps,
};
