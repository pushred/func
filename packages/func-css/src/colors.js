import chroma from 'chroma-js';
import isPlainObject from 'lodash.isplainobject';

function parse(colors = {}) {
  return Object.keys(colors).reduce((output, colorKey) => {
    const rawColor = colors[colorKey];
    if (typeof rawColor === 'string') output[colorKey] = chroma(rawColor);
    if (!isPlainObject(rawColor)) return output;

    let colorModelName = Object.keys(rawColor).map(val => val[0]).join('');
    if (colorModelName === 'hsb') colorModelName = 'hsv';

    const color = Object.keys(rawColor).map(key => (
      ['s', 'b', 'l', 'v'].includes(key[0]) && parseInt(rawColor[key], 10) > 1
        ? parseInt(rawColor[key], 10) / 100
        : rawColor[key]
    ));

    output[colorKey] = chroma.call(null, color, colorModelName);

    return output;
  }, {});
}

module.exports = {
  parse,
};
