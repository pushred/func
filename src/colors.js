import chroma from 'chroma-js';
import isPlainObject from 'lodash.isplainobject';

function parse(colors = {}) {
  return Object.keys(colors).reduce((output, colorKey) => {
    let colorVals = colors[colorKey];
    if (typeof colorVals === 'string') output[colorKey] = chroma(colorVals);
    if (!isPlainObject(colorVals)) return output;

    let colorModelName = Object.keys(colorVals).map(val => val[0]).join('');
    if (colorModelName === 'hsb') colorModelName = 'hsv';

    colorVals = Object.keys(colorVals).map(key => (
      ['s', 'b', 'l', 'v'].includes(key[0]) && parseInt(colorVals[key], 10) > 1
        ? parseInt(colorVals[key], 10) / 100
        : colorVals[key]
    ));

    output[colorKey] = chroma.call(null, colorVals, colorModelName);

    return output;
  }, {});
}

module.exports = {
  parse,
};
