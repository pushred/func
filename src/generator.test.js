import { expandClasses, generateClasses } from './generator';
import { parse } from './colors';

function testExpandClasses() {
  const colors = {
    red: '#ff0000',
    gray: '#fafafa',
    white: '#fff',
    blue: '#0000ff',
  };

  test('css keywords passthrough', () => {
    const classes = {
      'button': 'currentColor',
      'button-bg': 'inherit',
      'button-border': 'initial',
    };

    const output = expandClasses({ classes, colors });
    expect(output).toEqual(expect.stringContaining('.button { color: currentColor'));
    expect(output).toEqual(expect.stringContaining('.button-bg { background-color: inherit'));
    expect(output).toEqual(expect.stringContaining('.button-border { border-color: initial'));
  });

  test('assigns named color to specified classes', () => {
    const classes = {
      'button': 'white',
      'button-bg': 'red',
      'button-border': 'red',
    };

    const output = expandClasses({ classes, colors });
    expect(output).toEqual(expect.stringContaining('.button { color'));
    expect(output).toEqual(expect.stringContaining('.button-bg { background-color'));
    expect(output).toEqual(expect.stringContaining('.button-border { border-color'));
  });

  test('pseudo-classes', () => {
    const classes = {
      'button': 'white :link',
      'button-bg': 'red *:link *:visited',
      'button-border': 'red &:link',
    };

    const output = expandClasses({ classes, colors });
    expect(output).toEqual(expect.stringContaining('.button:link { color'));
    expect(output).toEqual(expect.stringContaining('.button-bg *:link { background-color'));
    expect(output).toEqual(expect.stringContaining('.button-bg *:visited { background-color'));
    expect(output).toEqual(expect.stringContaining('.button-border:link { border-color'));
  });

  test('pseudo-elements', () => {
    const classes = {
      'input': 'gray ::placeholder',
    };

    const output = expandClasses({ classes, colors });
    expect(output).toEqual(expect.stringContaining('.input::placeholder { color'));
  });

  test('alpha adjuster', () => {
    const classes = {
      'button': 'white a(0.8)',
      'button-bg': 'white alpha(90%)',
    };

    const output = expandClasses({ classes, colors });
    expect(output).toEqual(expect.stringContaining('hsla(0, 0%, 100%, 0.8)'));
    expect(output).toEqual(expect.stringContaining('hsla(0, 0%, 100%, 0.9)'));
  });
}

function testGenerateClasses() {
  const colors = {
    red: '#ff0000',
    blue: '#0000ff',
  };

  test('creates specified classes for each color', () => {
    const properties = {
      color: '${name}-color',
    };

    const output = generateClasses({ colors, properties });
    expect(output).toEqual(expect.stringContaining('.red-color'));
    expect(output).toEqual(expect.stringContaining('.blue-color'));
  });

  test('assigns color values in HSL by default', () => {
    const properties = {
      color: '${name}-color',
    };

    const output = generateClasses({ colors, properties });
    expect(output).toEqual(expect.stringContaining('hsl(0, 100%, 50%)'));
    expect(output).toEqual(expect.stringContaining('hsl(240, 100%, 50%)'));
  });

  test('pseudo-classes', () => {
    const config = {
      properties: {
        color: '${name}-color',
      },
      states: [
        'link',
      ],
    };

    const output = generateClasses({ ...config, colors });

    expect(output).toEqual(expect.stringContaining('.red-color:link'));
    expect(output).toEqual(expect.stringContaining('.blue-color:link'));
  });
}

describe('generator', () => {
  describe('expandClasses', testExpandClasses);
  describe('generateClasses', testGenerateClasses);
});
