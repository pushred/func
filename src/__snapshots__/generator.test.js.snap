// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generator expandClasses snapshot 1`] = `
".button { color: hsl(0, 0%, 100%) }
.button:link { color: hsl(0, 0%, 100%) }
.button:visited { color: hsl(0, 0%, 100%) }
.button:active { color: hsl(0, 0%, 100%) }
.button *:link { color: hsl(0, 0%, 100%) }
.button *:visited { color: hsl(0, 0%, 100%) }
.button *:active { color: hsl(0, 0%, 100%) }
.button-bg { background-color: hsl(0, 100%, 50%) }
.button-bg:link { background-color: hsl(0, 100%, 50%) }
.button-bg:visited { background-color: hsl(0, 100%, 50%) }
.button-bg:active { background-color: hsl(0, 100%, 50%) }
.button-bg-hover:hover { background-color: hsl(0, 100%, 50%) }
.button-bg-hover *:hover { background-color: hsl(0, 100%, 50%) }
.button-border { border-color: hsl(0, 100%, 50%) }
.button-border:link { border-color: hsl(0, 100%, 50%) }
.button-border:visited { border-color: hsl(0, 100%, 50%) }
.button-border:active { border-color: hsl(0, 100%, 50%) }
.input::placeholder { color: hsl(0, 0%, 98.04%) }"
`;

exports[`generator generateClasses snapshot 1`] = `
".red-color { color: hsl(0, 100%, 50%) }
.red-color:link { color: hsl(0, 100%, 50%) }
.red-color:visited { color: hsl(0, 100%, 50%) }
.red-color:active { color: hsl(0, 100%, 50%) }
.white-color { color: hsl(0, 0%, 100%) }
.white-color:link { color: hsl(0, 0%, 100%) }
.white-color:visited { color: hsl(0, 0%, 100%) }
.white-color:active { color: hsl(0, 0%, 100%) }
.blue-color { color: hsl(240, 100%, 50%) }
.blue-color:link { color: hsl(240, 100%, 50%) }
.blue-color:visited { color: hsl(240, 100%, 50%) }
.blue-color:active { color: hsl(240, 100%, 50%) }"
`;
